File: python/providers/langchain.py
import os
from posthog.ai.langchain import CallbackHandler
from langchain_openai import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.tools import tool
from langchain_core.messages import ToolMessage, HumanMessage, AIMessage, SystemMessage
from posthog import Posthog
from .base import BaseProvider
from .constants import OPENAI_CHAT_MODEL, OPENAI_VISION_MODEL, SYSTEM_PROMPT_ASSISTANT

class LangChainProvider(BaseProvider):
    def __init__(self, posthog_client: Posthog):
        super().__init__(posthog_client)
        self.OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
        
        self.callback_handler = CallbackHandler(
            client=posthog_client
        )
        
        # Store conversation history in LangChain's native format
        self.langchain_messages = [
            SystemMessage(content=SYSTEM_PROMPT_ASSISTANT)
        ]
        
        self._setup_chain()
    
    def get_tool_definitions(self):
        """Return tool definitions (not used by LangChain but required by base)"""
        return []
    
    def _setup_chain(self):
        """Setup the LangChain chain with tools"""
        @tool
        def get_weather(latitude: float, longitude: float, location_name: str) -> str:
            """Get the current weather for a specific location using geographical coordinates.

            Args:
                latitude: The latitude of the location (e.g., 37.7749 for San Francisco)
                longitude: The longitude of the location (e.g., -122.4194 for San Francisco)
                location_name: A human-readable name for the location (e.g., 'San Francisco, CA' or 'Dublin, Ireland')

            Returns:
                Weather information for the specified location
            """
            return self.get_weather(latitude, longitude, location_name)
        
        self.langchain_tools = [get_weather]
        self.tool_map = {tool.name: tool for tool in self.langchain_tools}
        
        prompt = ChatPromptTemplate.from_messages([
            ("system", SYSTEM_PROMPT_ASSISTANT),
            ("user", "{input}")
        ])
        
        model = ChatOpenAI(openai_api_key=self.OPENAI_API_KEY, temperature=0)
        self.chain = prompt | model.bind_tools(self.langchain_tools)
    
    def get_name(self):
        return "LangChain (OpenAI)"
    
    def get_description(self):
        return "💡 You can ask me about the weather!"
    
    def reset_conversation(self):
        """Reset the conversation history"""
        self.langchain_messages = [
            SystemMessage(content=SYSTEM_PROMPT_ASSISTANT)
        ]
        self.messages = []
    
    def chat(self, user_input: str, base64_image: str = None) -> str:
        """Send a message to LangChain and get response"""
        # Add user message to history
        if base64_image:
            # Create a message with image content
            user_message = HumanMessage(content=[
                {
                    "type": "text",
                    "text": user_input
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/png;base64,{base64_image}"
                    }
                }
            ])
        else:
            user_message = HumanMessage(content=user_input)
        self.langchain_messages.append(user_message)
        
        # Use the model directly with conversation history instead of the chain
        # Use vision model for images
        model_name = OPENAI_VISION_MODEL if base64_image else OPENAI_CHAT_MODEL
        model = ChatOpenAI(openai_api_key=self.OPENAI_API_KEY, temperature=0, model_name=model_name)
        model_with_tools = model.bind_tools(self.langchain_tools)

        # Prepare API request parameters
        request_params = {
            "input": self.langchain_messages,
            "config": {"callbacks": [self.callback_handler]}
        }

        response = model_with_tools.invoke(**request_params)

        # Debug: Log the API call (request + response)
        self._debug_api_call("LangChain (OpenAI)", request_params, response)

        # Collect display parts
        display_parts = []
        
        # Add the AI's text response to display (if any)
        if response.content:
            display_parts.append(response.content)
        
        # Check if the model wants to use tools
        if response.tool_calls:
            # Execute tool calls
            tool_messages = []
            for tool_call in response.tool_calls:
                tool_name = tool_call["name"]
                tool_args = tool_call["args"]
                
                # Execute the tool
                if tool_name in self.tool_map:
                    tool_result = self.tool_map[tool_name].invoke(tool_args)
                    tool_result_text = self.format_tool_result("get_weather", tool_result)
                    display_parts.append(tool_result_text)
                    
                    tool_messages.append(
                        ToolMessage(
                            content=str(tool_result),
                            tool_call_id=tool_call["id"],
                        )
                    )
        
        # Add assistant's response to conversation history
        self.langchain_messages.append(response)
        
        # Add tool messages to history if any
        if response.tool_calls:
            self.langchain_messages.extend(tool_messages)
        
        return "\n\n".join(display_parts) if display_parts else "No response received"

File: node/src/providers/langchain.ts
import { LangChainCallbackHandler } from '@posthog/ai';
import { ChatOpenAI } from '@langchain/openai';
import { ChatPromptTemplate } from '@langchain/core/prompts';
import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { HumanMessage, SystemMessage, ToolMessage } from '@langchain/core/messages';
import { PostHog } from 'posthog-node';
import { BaseProvider, Tool } from './base.js';
import { OPENAI_CHAT_MODEL, OPENAI_VISION_MODEL, SYSTEM_PROMPT_ASSISTANT } from './constants.js';

export class LangChainProvider extends BaseProvider {
  private callbackHandler: any;
  private langchainMessages: any[] = [];
  private langchainTools: any[] = [];
  private toolMap: Map<string, any> = new Map();

  constructor(posthogClient: PostHog, aiSessionId: string | null = null) {
    super(posthogClient, aiSessionId);
    this.callbackHandler = new LangChainCallbackHandler({
      client: posthogClient,
      properties: {
        $ai_span_name: "langchain_chat",
        ...this.getPostHogProperties(),
      },
    });

    this.langchainMessages = [
      new SystemMessage(SYSTEM_PROMPT_ASSISTANT)
    ];

    this.setupChain();
  }

  protected getToolDefinitions(): Tool[] {
    // LangChain doesn't use this, but we need to implement it
    return [];
  }

  private setupChain(): void {
    const getWeatherTool = new DynamicStructuredTool({
      name: 'get_weather',
      description: 'Get the current weather for a specific location',
      schema: z.object({
        latitude: z.number().describe('The latitude of the location (e.g., 37.7749 for San Francisco)'),
        longitude: z.number().describe('The longitude of the location (e.g., -122.4194 for San Francisco)'),
        location_name: z.string().describe('A human-readable name for the location (e.g., \'San Francisco, CA\' or \'Dublin, Ireland\')'),
      }),
      func: async (input: { latitude: number; longitude: number; location_name: string }) => {
        return this.getWeather(input.latitude, input.longitude, input.location_name);
      }
    } as any) as any;

    this.langchainTools = [getWeatherTool];
    this.toolMap.set('get_weather', getWeatherTool);

    const prompt = ChatPromptTemplate.fromMessages([
      ['system', SYSTEM_PROMPT_ASSISTANT],
      ['user', '{input}']
    ]);

    const model = new ChatOpenAI({ openAIApiKey: process.env.OPENAI_API_KEY, temperature: 0 });
    prompt.pipe(model.bindTools(this.langchainTools));
  }

  getName(): string {
    return 'LangChain (OpenAI)';
  }

  getDescription(): string {
    return '💡 You can ask me about the weather!';
  }

  resetConversation(): void {
    this.langchainMessages = [
      new SystemMessage(SYSTEM_PROMPT_ASSISTANT)
    ];
    this.messages = [];
  }

  async chat(userInput: string, base64Image?: string): Promise<string> {
    let userMessage: any;
    
    if (base64Image) {
      // Create a message with image content
      userMessage = new HumanMessage({
        content: [
          {
            type: 'text',
            text: userInput
          },
          {
            type: 'image_url',
            image_url: {
              url: `data:image/png;base64,${base64Image}`
            }
          }
        ]
      });
    } else {
      userMessage = new HumanMessage(userInput);
    }
    
    this.langchainMessages.push(userMessage);

    const model = new ChatOpenAI({
      openAIApiKey: process.env.OPENAI_API_KEY,
      temperature: 0,
      modelName: base64Image ? OPENAI_VISION_MODEL : OPENAI_CHAT_MODEL
    });
    const modelWithTools = model.bindTools(this.langchainTools);

    const requestParams = {
      messages: this.langchainMessages,
      callbacks: [this.callbackHandler]
    };

    const response = await modelWithTools.invoke(
      this.langchainMessages,
      { callbacks: [this.callbackHandler] }
    );
    this.debugApiCall("LangChain (OpenAI)", requestParams, response);

    const displayParts: string[] = [];

    if (response.content) {
      displayParts.push(response.content as string);
    }

    if (response.tool_calls && response.tool_calls.length > 0) {
      const toolMessages: any[] = [];
      for (const toolCall of response.tool_calls) {
        const toolName = toolCall.name;
        const toolArgs = toolCall.args;

        if (this.toolMap.has(toolName)) {
          const toolResult = await this.toolMap.get(toolName)!.invoke(toolArgs);
          const toolResultText = this.formatToolResult('get_weather', toolResult);
          displayParts.push(toolResultText);

          toolMessages.push(
            new ToolMessage({
              content: String(toolResult),
              tool_call_id: toolCall.id!,
            })
          );
        }
      }

      this.langchainMessages.push(response);
      this.langchainMessages.push(...toolMessages);
    } else {
      this.langchainMessages.push(response);
    }

    return displayParts.length > 0 ? displayParts.join('\n\n') : 'No response received';
  }
}